### ISSUE TITLE
Error building elpa: AttributeError: 'Spec' object has no attribute 'mpicc'

### ISSUE THREAD

**Issue by baip** [spack/spack#6143]
I was trying to do: `spack install elpa %gcc@6.4.0 ^intel-mkl@2018.0.128 ^intel-mpi@2018.0.128` on Rocks cluster 6.2/CentOS 6.6, but in the install phase, it gives the error in the title. The spack-build.out indicates however that the build was successful:

    ----------------------------------------------------------------------
    Libraries have been installed in:
       /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/lib

    If you ever happen to want to link against installed libraries
    in a given directory, LIBDIR, you must either use libtool, and
    specify the full pathname of the library, or use the '-LLIBDIR'
    flag during linking and do at least one of the following:
      - add LIBDIR to the 'LD_LIBRARY_PATH' environment variable
        during execution
      - add LIBDIR to the 'LD_RUN_PATH' environment variable
        during linking
      - use the '-Wl,-rpath -Wl,LIBDIR' linker flag
      - have your system administrator add LIBDIR to '/etc/ld.so.conf'
    
    See any operating system documentation about shared libraries for
    more information, such as the ld(1) and ld.so(8) manual pages.
     ----------------------------------------------------------------------
    /bin/mkdir -p '/share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/bin'
     /bin/sh ./libtool   --mode=install /usr/bin/install -c elpa2_print_kernels '/share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/bin'
    libtool: install: /usr/bin/install -c .libs/elpa2_print_kernels /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-7.2.0/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/bin/elpa2_print_kernels
    make[2]: Leaving directory `/tmp/soft/spack-stage/spack-stage-aaRYV2/elpa-2016.05.004/spack-build'
    make[1]: Leaving directory `/tmp/soft/spack-stage/spack-stage-aaRYV2/elpa-2016.05.004/spack-build'

The full error logs are:

    ==> intel-mkl is already installed in /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/intel-mkl-2018.0.128-ghtmbtscrmcextiiwijcgh2pv2xftq6r                                       
    ==> intel-mpi is already installed in /share/apps/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/intel-mpi-2018.0.128-hfjbyb3cljw5ej2xtil6vujvjrvywbc2                                       
    ==> Installing elpa                                                                                                                                                                          
    ('spec-mpi', ['__class__', '__contains__', '__delattr__', '__dict__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_add_dependency', '_add_flag', '_add_version', '_autospec', '_cmp_key', '_cmp_key_cache', '_cmp_node', '_concrete', '_concretize_helper', '_constrain_dependencies', '_dependencies', '_dependents', '_dup', '_dup_deps', '_eq_dag', '_evaluate_dependency_conditions', '_expand_virtual_packages', '_find_deps', '_find_provider', '_hash', '_install_status', '_installed_explicitly', '_mark_concrete', '_merge_dependency', '_normal', '_normalize_helper', '_replace_with', '_set_architecture', '_set_compiler', '_set_test_prefix', 'architecture', 'cformat', 'colorized', 'command', 'common_dependencies', 'compiler', 'compiler_flags', 'concrete', 'concretize', 'concretized', 'constrain', 'constrained', 'copy', 'cshort_spec', 'dag_hash', 'dag_hash_bit_prefix', 'dep_difference', 'dep_string', 'dependencies', 'dependencies_dict', 'dependents', 'dependents_dict', 'eq_dag', 'eq_node', 'external', 'external_module', 'external_path', 'flat_dependencies', 'format', 'from_dict', 'from_json', 'from_literal', 'from_node_dict', 'from_yaml', 'fullname', 'get_dependency', 'headers', 'index', 'is_virtual', 'last_query', 'libs', 'mpicc', 'mpicxx', 'mpif77', 'mpifc', 'name', 'namespace', 'ne_dag', 'ne_node', 'normalize', 'normalized', 'package', 'package_class', 'patches', 'prefix', 'read_yaml_dep_specs', 'root', 'satisfies', 'satisfies_dependencies', 'short_spec', 'sorted_deps', 'to_dict', 'to_json', 'to_node_dict', 'to_yaml', 'traverse', 'traverse_edges', 'tree', 'validate_or_raise', 'variants', 'version', 'versions', 'virtual', 'virtual_dependencies'])
    ==> Warning: Suspicious requests to set or unset 'CC' found
    ==> Warning:            env.set('CC', join_path(link_dir, compiler.link_paths['cc'])) at /share/apps/spack/lib/spack/spack/build_environment.py:148
    ==> Warning:    --->    spack_env.set('CC', spec['mpi'].mpicc) at /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:72
    ==> Warning: Suspicious requests to set or unset 'CXX' found
    ==> Warning:            env.set('CXX', join_path(link_dir, compiler.link_paths['cxx'])) at /share/apps/spack/lib/spack/spack/build_environment.py:151
    ==> Warning:    --->    spack_env.set('CXX', spec['mpi'].mpicxx) at /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:74
    ==> Warning: Suspicious requests to set or unset 'FC' found
    ==> Warning:            env.set('FC', join_path(link_dir, compiler.link_paths['fc'])) at /share/apps/spack/lib/spack/spack/build_environment.py:157
    ==> Warning:    --->    spack_env.set('FC', spec['mpi'].mpifc) at /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:73
    ==> Using cached archive: /share/apps/spack/var/spack/cache/elpa/elpa-2016.05.004.tar.gz
    ==> Staging archive: /share/apps/spack/var/spack/stage/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj/elpa-2016.05.004.tar.gz
    ==> Created stage in /share/apps/spack/var/spack/stage/elpa-2016.05.004-7r6rtnnke4rasnbmww6thzgdv57b2nmj
    ==> No patches needed for elpa
    ==> Building elpa [AutotoolsPackage]
    ==> Executing phase: 'autoreconf'
    ==> Executing phase: 'configure'
    ==> Executing phase: 'build'
    ==> Executing phase: 'install'
    ('spec-mpi', ['__class__', '__contains__', '__delattr__', '__dict__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__gt__', '__hash__', '__init__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_add_dependency', '_add_flag', '_add_version', '_autospec', '_cmp_key', '_cmp_key_cache', '_cmp_node', '_concrete', '_concretize_helper', '_constrain_dependencies', '_dependencies', '_dependents', '_dup', '_dup_deps', '_eq_dag', '_evaluate_dependency_conditions', '_expand_virtual_packages', '_find_deps', '_find_provider', '_hash', '_install_status', '_installed_explicitly', '_mark_concrete', '_merge_dependency', '_normal', '_normalize_helper', '_replace_with', '_set_architecture', '_set_compiler', '_set_test_prefix', 'architecture', 'cformat', 'colorized', 'command', 'common_dependencies', 'compiler', 'compiler_flags', 'concrete', 'concretize', 'concretized', 'constrain', 'constrained', 'copy', 'cshort_spec', 'dag_hash', 'dag_hash_bit_prefix', 'dep_difference', 'dep_string', 'dependencies', 'dependencies_dict', 'dependents', 'dependents_dict', 'eq_dag', 'eq_node', 'external', 'external_module', 'external_path', 'flat_dependencies', 'format', 'from_dict', 'from_json', 'from_literal', 'from_node_dict', 'from_yaml', 'fullname', 'get_dependency', 'headers', 'index', 'is_virtual', 'last_query', 'libs', 'name', 'namespace', 'ne_dag', 'ne_node', 'normalize', 'normalized', 'package', 'package_class', 'patches', 'prefix', 'read_yaml_dep_specs', 'root', 'satisfies', 'satisfies_dependencies', 'short_spec', 'sorted_deps', 'to_dict', 'to_json', 'to_node_dict', 'to_yaml', 'traverse', 'traverse_edges', 'tree', 'validate_or_raise', 'variants', 'version', 'versions', 'virtual', 'virtual_dependencies'])
    ==> Error: AttributeError: 'Spec' object has no attribute 'mpicc'

    /share/apps/spack/var/spack/repos/builtin/packages/elpa/package.py:71, in setup_environment:
         3             spec = self.spec
         4
         5             print('spec-mpi', dir(spec['mpi']))
      >> 6             spack_env.set('CC', spec['mpi'].mpicc)
         7             spack_env.set('FC', spec['mpi'].mpifc)
         8             spack_env.set('CXX', spec['mpi'].mpicxx)

**Comment by pramodskumbhar** [comment]
I haven't used elpa package but with following error:

```
==> Error: AttributeError: 'Spec' object has no attribute 'mpicc'
AttributeError: AttributeError: 'Spec' object has no attribute 'mpicc'
```
I wonder about your mpi installation. Is this Spack installed intel mpi? Or you have configured it as external package in packages.yaml?

**Comment by baip** [comment]
I used Spack to install intel-mpi & intel-mkl. Note that the package was apparently built successfully, and `setup_environment()` in `elpa/package.py` was called twice, only the second time during the install phase produced the error.

**Comment by pramodskumbhar** [comment]
> Note that the package was apparently built successfully, and setup_environment() in elpa/package.py was called twice

The function setup_environment is called twice : install time to set up the build environment and after the install to generate the associated module file.

**Comment by baip** [comment]
Yes. It's buried in the attached log, but the error happens the second time it was called, after `==> Executing phase: 'install'`

**Comment by tgamblin** [comment]
@baip: any chance you can run this again with `spack -d` so that the full stacktrace is printed?

If I just look at the `elpa` package, this is the only place `mpicc` is accessed:

```python
    def setup_environment(self, spack_env, run_env):                                                                    
        # TODO: set optimum flags for platform+compiler combo, see                                                      
        # https://github.com/hfp/xconfigure/tree/master/elpa                                                            
        spec = self.spec                                                                                                
                                                                                                                        
        spack_env.set('CC', spec['mpi'].mpicc)                                                                          
        spack_env.set('FC', spec['mpi'].mpifc)                                                                          
        spack_env.set('CXX', spec['mpi'].mpicxx)  
[...]
```

But it's set on the `intel-mpi` spec in `intel-mpi`'s `setup_dependent_package` method:

```python
    def setup_dependent_package(self, module, dep_spec):                                                                
[...]
        if self.compiler.name == 'intel':                                                                               
            self.spec.mpicc  = bindir.mpiicc                                                                            
            self.spec.mpicxx = bindir.mpiicpc                                                                           
            self.spec.mpifc  = bindir.mpiifort                                                                          
            self.spec.mpif77 = bindir.mpiifort                                                                          
        else:                                                                                                           
            self.spec.mpicc  = bindir.mpicc                                                                             
            self.spec.mpicxx = bindir.mpicxx                                                                            
            self.spec.mpifc  = bindir.mpif90                                                                            
            self.spec.mpif77 = bindir.mpif77     
```

This is called *before* `elpa`'s `setup_environment` so it should be accessible there.  What is `spec['mpi']` in that method?  Is it the `intel-mpi` spec?

**Comment by baip** [comment]
@tgamblin You're right. The error occurred in `elpa`'s `setup_environment`, which was called twice, and only the second time `spec['mpi']` is missing. Here is the full debug log:

    ==> Reading config file /panfs/roc/groups/1/group/shared/tools/spack/etc/spack/defaults/packages.yaml
    ==> Reading config file /panfs/roc/groups/1/group/shared/tools/spack/etc/spack/packages.yaml
    ==> READ LOCK: /home/group/user/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquiring]
    ==> READ LOCK: /home/group/user/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquired]
    ==> READ LOCK: /home/group/user/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Released]
    ==> Reading config file /panfs/roc/groups/1/group/shared/tools/spack/etc/spack/compilers.yaml
    ==> Reading config file /home/group/user/.spack/linux/compilers.yaml
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Acquiring]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Acquired]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/lock[0:0] [Released]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Released]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Acquiring]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Acquired]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Released]
    ==> Installing elpa dependencies
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[1797886672307061285:1] [Acquiring]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[1797886672307061285:1] [Acquired]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[1797886672307061285:1] [Released]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[1797886672307061285:1] [Acquiring]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[1797886672307061285:1] [Acquired]
    ==> intel-mkl is already installed in /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/intel-mkl-2018.0.128-ghtmbtscrmcextiiwijcgh2pv2xftq6r
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[1797886672307061285:1] [Released]
    ==> intel-mpi@2018.0.128 : externally installed in /panfs/roc/intel/x86_64/2018/impi_msi
    ==> intel-mpi@2018.0.128 : already registered in DB
    ==> Installing elpa
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> Warning: Suspicious requests to set or unset 'CC' found
    ==> Warning:            env.set('CC', join_path(link_dir, compiler.link_paths['cc'])) at /panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/build_environment.py:146
    ==> Warning:    --->    spack_env.set('CC', spec['mpi'].mpicc) at /panfs/roc/groups/1/group/shared/tools/spack/var/spack/repos/builtin/packages/elpa/package.py:70
    ==> Warning: Suspicious requests to set or unset 'CXX' found
    ==> Warning:            env.set('CXX', join_path(link_dir, compiler.link_paths['cxx'])) at /panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/build_environment.py:149
    ==> Warning:    --->    spack_env.set('CXX', spec['mpi'].mpicxx) at /panfs/roc/groups/1/group/shared/tools/spack/var/spack/repos/builtin/packages/elpa/package.py:72
    ==> Warning: Suspicious requests to set or unset 'FC' found
    ==> Warning:            env.set('FC', join_path(link_dir, compiler.link_paths['fc'])) at /panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/build_environment.py:155
    ==> Warning:    --->    spack_env.set('FC', spec['mpi'].mpifc) at /panfs/roc/groups/1/group/shared/tools/spack/var/spack/repos/builtin/packages/elpa/package.py:71
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'binutils/2.28'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'load' 'binutils/2.28'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'load' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> Using cached archive: /panfs/roc/groups/1/group/shared/tools/spack/var/spack/cache/elpa/elpa-2016.05.004.tar.gz
    ==> Already staged elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu in /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu
    ==> No patches needed for elpa
    ==> Building elpa [AutotoolsPackage]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/.lock[1697835973088342236:1] [Acquiring]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/.lock[1697835973088342236:1] [Acquired]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Acquiring]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Acquired]
    ==> Executing phase: 'autoreconf'
    ==> Executing phase: 'configure'
    ==> Executing phase: 'build'
    ==> Executing phase: 'install'
    ==> Installing /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/elpa-2016.05.004/spack-build.out to /panfs/roc/groups/1/neur
    ockm/shared/tools/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/.spack/build.out
    ==> Installing /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/elpa-2016.05.004/spack-build.env to /panfs/roc/groups/1/neur
    ockm/shared/tools/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/.spack/build.env
    ==> Installing /panfs/roc/groups/1/group/shared/tools/spack/var/spack/repos/builtin/packages/elpa/package.py to /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/linux-centos6-x8
    6_64/gcc-6.4.0/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/.spack/repos/builtin/packages/elpa
    ==> Installing /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/intel-mkl-2018.0.128-ghtmbtscrmcextiiwijcgh2pv2xftq6r/.spack/repos/builtin/packages/intel-mkl to /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/linux-centos6-x86_64/gcc-6.4.0/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/.spack/repos/builtin/packages/intel-mkl
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
    ==> READ LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/lock[0:0] [Released]
    ==>     WRITE: elpa@2016.05.004%gcc@6.4.0~openmp arch=linux-centos6-x86_64 /zdz4gag [/panfs/roc/groups/1/group/shared/tools/spack/share/spack/modules/linux-centos6-x86_64/elpa-2016.05.004-gcc-6.4.0-zdz4gag]
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'list'
    ==> '/soft/Modules/3.2.6/bin/modulecmd' 'python' 'show' 'gcc-6.4.0-gcc-4.4.7-bx5dbff'
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/opt/spack/.spack-db/prefix_lock[7240065817500364352:1] [Released]
    ==> WRITE LOCK: /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/.lock[1697835973088342236:1] [Released]
    ==> Error: AttributeError: 'Spec' object has no attribute 'mpicc'
    AttributeError: AttributeError: 'Spec' object has no attribute 'mpicc'
    
    /panfs/roc/groups/1/group/shared/tools/spack/var/spack/repos/builtin/packages/elpa/package.py:70, in setup_environment:
         2             # https://github.com/hfp/xconfigure/tree/master/elpa
         3             spec = self.spec
         4
      >> 5             spack_env.set('CC', spec['mpi'].mpicc)
         6             spack_env.set('FC', spec['mpi'].mpifc)
         7             spack_env.set('CXX', spec['mpi'].mpicxx)
         8
    
    See build log for details:
      /panfs/roc/groups/1/group/shared/tools/spack/var/spack/stage/elpa-2016.05.004-zdz4gagfdvoicx6dmakb6wjlubd66bvu/elpa-2016.05.004/spack-build.out
    Traceback (most recent call last):
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/build_environment.py", line 662, in child_process
        return_value = function()
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/package.py", line 1447, in build_process
        spack.hooks.post_install(self.spec)
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/hooks/__init__.py", line 73, in __call__
        hook(*args, **kwargs)
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/hooks/module_file_generation.py", line 48, in <lambda>
        post_install = lambda spec: _for_each_enabled(spec, 'write')
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/hooks/module_file_generation.py", line 41, in _for_each_enabled
        getattr(generator, method_name)()
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/modules/common.py", line 697, in write
        context = self.context.to_dict()
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/tengine.py", line 85, in to_dict
        d = [(name, getattr(self, name)) for name in self.context_properties]
      File "/panfs/roc/groups/1/group/shared/tools/spack/lib/spack/spack/modules/common.py", line 552, in environment_modifications
        self.spec.package.setup_environment(_, env)
      File "/panfs/roc/groups/1/group/shared/tools/spack/var/spack/repos/builtin/packages/elpa/package.py", line 70, in setup_environment
        spack_env.set('CC', spec['mpi'].mpicc)
    AttributeError: 'Spec' object has no attribute 'mpicc'

**Comment by tgamblin** [comment]
@alalazo: it looks like something is wrong with the way the module hook is aggregating environment modifications.  I'm not familiar with that part of the code, but it looks like it's reimplementing some a lot of what `build_environment` is doing.  Oh.  It looks like you know that:

```python
        # TODO : the code down below is quite similar to                                                                
        # TODO : build_environment.setup_package and needs to be factored out                                           
        # TODO : to a single place   
```

😄 

Would it make sense to replace the code in `modules/common.py` with a call to `build_environment.fork()` with some custom function?  You can return a value from the forked function, so you could access the build environment and get back what you need.

**Comment by alalazo** [comment]
> it looks like something is wrong with the way the module hook is aggregating environment modifications. I'm not familiar with that part of the code, but it looks like it's reimplementing some a lot of what build_environment is doing

Yes, that dates back to #552 and probably needs some rethinking. We don't want this for SC17, right? Currently I am trying to avoid modifications to core, so that we don't introduce further bugs.

> Would it make sense to replace the code in modules/common.py with a call to build_environment.fork() with some custom function? You can return a value from the forked function, so you could access the build environment and get back what you need.

@tgamblin Not sure about that, need to have a look at the implications. What I am thinking since a while is that it definitely makes sense to separate the callbacks for build and run-time environment. What I mean is going from this:
```python
def setup_environment(self, spack_env, run_env):                                                                    
    ...
```
to this:
```python
def setup_build_environment(self, env):                                                                    
    ...

def setup_run_environment(self, env):                                                                    
    ...
```
The rationale is that we call `setup_environment` 2 times right now, and we discard one time  `run_env`, the other time `spack_env` + most packages either add changes to one or the other variable. What do you think?

**Comment by alalazo** [comment]
Relates to #5864

**Comment by alalazo** [comment]
I think this has been fixed by #7469
